<!DOCTYPE html>
<html lang="en"><head>
<script src="factor_analysis_files/libs/clipboard/clipboard.min.js"></script>
<script src="factor_analysis_files/libs/quarto-html/tabby.min.js"></script>
<script src="factor_analysis_files/libs/quarto-html/popper.min.js"></script>
<script src="factor_analysis_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="factor_analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="factor_analysis_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="factor_analysis_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="factor_analysis_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.5.57">

  <meta name="author" content="Dr.&nbsp;Jyotishka Datta">
  <title>Factor Analysis   Introduction</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="factor_analysis_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="factor_analysis_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="factor_analysis_files/libs/revealjs/dist/theme/quarto.css">
  <link href="factor_analysis_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="factor_analysis_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="factor_analysis_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="factor_analysis_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Factor Analysis <br> Introduction</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Dr.&nbsp;Jyotishka Datta 
</div>
</div>
</div>

</section>
<section>
<section id="factor-analysis" class="title-slide slide level1 center">
<h1>Factor Analysis</h1>

</section>
<section id="source" class="slide level2">
<h2>Source</h2>
<ul>
<li><p>We will briefly talk about exploratory factor analysis or EFA, and defer the confirmatory factor analysis (CFA) for a later date, if time permits.</p></li>
<li><p>These slides are (mostly) based on chapter 8 of Zelterman and chapter 5 of Everitt.</p></li>
<li><p>Everitt is much more detailed: two chapters for EFA and CFA.</p></li>
</ul>
</section>
<section id="pca-and-fa" class="slide level2">
<h2>PCA and FA</h2>
<ul>
<li><p>FA is viewed as being different from PCA</p></li>
<li><p><strong>PCA</strong></p></li>
<li><p>is more exploratory</p></li>
<li><p>may not require as many assumptions</p></li>
<li><p><strong>FA</strong></p></li>
<li><p>may require additional assumptions (normality)</p></li>
<li><p>often associated with a search for causal models</p></li>
<li><p>rotation is important</p></li>
<li><p><strong>Big difference is philosophical</strong></p></li>
<li><p>FA program will do a PCA but not visa versa</p></li>
</ul>
</section>
<section id="motivation" class="slide level2">
<h2>Motivation</h2>
<p>In many areas of psychology, and other disciplines in the behavioural sciences, often it is not possible to measure directly the concepts of primary interest.</p>
<p>Two obvious examples are intelligence and social class. In such cases, the researcher is forced to examine the concepts indirectly by collecting information on variables that can be measured or observed directly and can also realistically be assumed to be indicators.</p>
<p>The psychologist who is interested in an individual’s “intelligence”, for example, may record examination scores in a variety of different subjects in the expectation that these scores are dependent in some way on what is widely regarded as “intelligence” but are also subject to random errors.</p>
<p>And a sociologist, say, concerned with people’s “social class” might pose questions about a person’s occupation, educational background, home ownership, etc., on the assumption that these do reflect the concept he or she is really interested in.</p>
</section>
<section id="latent-variables" class="slide level2">
<h2>Latent variables</h2>
<p>Both “intelligence” and “social class” are what are generally referred to as <strong>latent variables</strong></p>
<p><strong>latent variables</strong> concepts that cannot be measured directly but can be assumed to relate to a number of measurable or <strong>manifest variables</strong>.</p>
<p>The method of analysis most generally used to help uncover the relationships between the assumed latent variables and the manifest variables is <strong>factor analysis.</strong></p>
<p>The model on which the method is based is essentially that of <strong>multiple regression,</strong> except now the manifest variables are regressed on the unobservable latent variables (often referred to in this context as <strong>common factors</strong>), so that direct estimation of the corresponding regression coeffcients (<strong>factor loadings</strong>) is not possible.</p>
</section>
<section id="types-of-factor-analysis" class="slide level2">
<h2>Types of factor analysis</h2>
<p>The first is <strong>exploratory factor analysis</strong>, which is used to investigate the relationship between manifest variables and factors without making any assumptions about which manifest variables are related to which factors.</p>
<p>The second is <strong>confirmatory factor analysis</strong> which is used to test whether a specific factor model postulated a priori provides an adequate fit for the covariances or correlations between the manifest variables.</p>
<p>We will first consider the EFA.</p>
</section>
<section id="simple-example" class="slide level2">
<h2>Simple example</h2>
<p>Spearman considered a sample of children’s examination marks in three subjects, Classics (<span class="math inline">\(y_1\)</span>), French (<span class="math inline">\(y_2\)</span>), and English (<span class="math inline">\(y_3\)</span>), from which he calculated the following correlation matrix for a sample of children:</p>
<p><span class="math display">\[
\mathbf{R} =
\begin{bmatrix}
    1.00 &amp;  &amp;  \\
    0.83 &amp; 1.00 &amp;  \\
    0.78 &amp; 0.67 &amp; 1.00
\end{bmatrix}.
\]</span></p>
<p>If we assume a single factor, then the <em>single-factor model</em> is specified as follows:</p>
<p><span class="math display">\[
\begin{aligned}
    y_1 &amp;= \lambda_1 f + u_1, \\
    y_2 &amp;= \lambda_2 f + u_2, \\
    y_3 &amp;= \lambda_3 f + u_3.
\end{aligned}
\]</span> In this example, the underlying latent variable or common factor, <span class="math inline">\(f\)</span>, might possibly be equated with intelligence or general intellectual ability.</p>
</section>
<section id="explanation" class="slide level2">
<h2>Explanation</h2>
<p>The terms <span class="math inline">\(\lambda_1, \lambda_2\)</span> and <span class="math inline">\(\lambda_3\)</span> which are essentially regression coeffcients are, in this context, known as factor loadings.</p>
<p>The terms <span class="math inline">\(u_1\)</span>, <span class="math inline">\(u_2\)</span>, and <span class="math inline">\(u_3\)</span> represent random disturbance terms and will have small variances if their associated observed variable is closely related to the underlying latent variable.</p>
<p>The variation in <span class="math inline">\(u_i\)</span> actually consists of two parts, the extent to which an individual’s ability at Classics, say, differs from his or her general ability and the extent to which the examination in Classics is only an approximate measure of his or her ability in the subject.</p>
<p>In practise, no attempt is made to disentangle these two parts.</p>
</section>
<section id="the-k-factor-analysis-model" class="slide level2">
<h2>The <span class="math inline">\(k\)</span>-factor analysis model</h2>
<p>To begin, we assume that we have a set of observed or manifest variables,<br>
<span class="math inline">\(\mathbf{y}^{\top} = (y_1, y_2, \dots, y_p)\)</span>, assumed to be linked to <span class="math inline">\(k\)</span> unobserved latent variables<br>
or common factors <span class="math inline">\(f_1, f_2, \dots, f_m\)</span>, where <span class="math inline">\(m &lt; p\)</span>, by a regression model of the form</p>
<p><span class="math display">\[
\begin{aligned}
    y_1 &amp;= \lambda_{11} f_1 + \lambda_{12} f_2 + \dots + \lambda_{1m} f_m + u_1, \\
    y_2 &amp;= \lambda_{21} f_1 + \lambda_{22} f_2 + \dots + \lambda_{2m} f_m + u_2, \\
    &amp;\vdots \\
    y_p &amp;= \lambda_{p1} f_1 + \lambda_{p2} f_2 + \dots + \lambda_{pm} f_m + u_p.
\end{aligned}
\]</span></p>
<p>The <span class="math inline">\(\lambda_j\)</span>s are essentially the regression coefficients of the <span class="math inline">\(y\)</span>-variables on the<br>
common factors.</p>
</section>
<section id="the-k-factor-analysis-model-1" class="slide level2">
<h2>The <span class="math inline">\(k\)</span>-factor analysis model</h2>
<p>… but in the context of factor analysis these regression coefficients are known as the factor loadings and show how each observed variable, <span class="math inline">\(y_i\)</span>, depends on the common factors.</p>
<p>The factor loadings are used in the interpretation of the factors; i.e., larger values relate a factor to the corresponding observed variables</p>
<p>… and from these we can often, but not always, infer a meaningful description of each factor.</p>
</section>
<section id="the-k-factor-analysis-model-2" class="slide level2">
<h2>The <span class="math inline">\(k\)</span>-factor analysis model</h2>
<p>The regression equations above may be written more concisely as</p>
<p><span class="math display">\[
\underbrace{\mathbf{y - \mu}}_{p \times 1} = \underbrace{\mathbf{\Lambda}}_{p \times m} ~~ \underbrace{\mathbf{f}}_{m \times 1} + \underbrace{\mathbf{u}}_{p \times 1}. \tag{F1}
\]</span></p>
<p>where</p>
<p><span class="math display">\[
\mathbf{\Lambda} =
\begin{pmatrix}
    \lambda_{11} &amp; \dots &amp; \lambda_{1m} \\
    \vdots &amp; \ddots &amp; \vdots \\
    \lambda_{p1} &amp; \dots &amp; \lambda_{pm}
\end{pmatrix},
\quad
\mathbf{f} =
\begin{pmatrix}
    f_1 \\
    \vdots \\
    f_m
\end{pmatrix},
\quad
\mathbf{u} =
\begin{pmatrix}
    u_1 \\
    \vdots \\
    u_p
\end{pmatrix}.
\]</span></p>
<p>We assume that 1. the random disturbance terms <span class="math inline">\(u_1, \dots, u_p\)</span> are uncorrelated with each other, and 2. uncorrelated with the factors <span class="math inline">\(f_1, \dots, f_m\)</span>.</p>
<p>(The elements of <span class="math inline">\(\mathbf{u}\)</span> are specific to each <span class="math inline">\(y_i\)</span> and hence are generally better known in this context as <em>specific variates</em>.)</p>
</section>
<section id="assumptions" class="slide level2">
<h2>Assumptions</h2>
<p>Specifically, we need to assume:</p>
<p><span class="math display">\[
\mathbb{E} f = \mathbf{0} \quad \text{and} \quad \operatorname{Cov}(f) = \mathbb{E}\{f f'\} = \mathbf{I}, \tag{A1}
\]</span></p>
<p><span class="math display">\[
\mathbb{E} \mathbf{u} = \mathbf{0} \quad \text{and} \quad \operatorname{Cov}(\mathbf{u}) = \mathbb{E}(\mathbf{u} \mathbf{u}') = \mathbf{\Psi} = \operatorname{Diag}(\Psi_1, \Psi_2, \dots, \Psi_p).\tag{A2}
\]</span></p>
<p>In addition, the factors ( f ) are uncorrelated with the errors ( u ):</p>
<p><span class="math display">\[
\operatorname{Cov}(f, \mathbf{u}) = \mathbf{0}.\tag{A3}
\]</span></p>
<p>Together, these assumptions (A1-A3) are sufficient for us to fit the model.</p>
</section>
<section id="the-k-factor-analysis-model-3" class="slide level2">
<h2>The <span class="math inline">\(k\)</span>-factor analysis model</h2>
<p>The two assumptions imply that, given the values of the common factors <span class="math inline">\(f\)</span>, the manifest variables <span class="math inline">\(y\)</span> are independent; that is, the correlations of the observed variables arise from their relationships with the common factors.</p>
<p>We also assume that factors are uncorrelated with one another, in which case the factor loadings are the correlations of the manifest variables and the factors.</p>
</section>
<section id="model" class="slide level2">
<h2>Model</h2>
<p>These assumptions can also be used to describe an alternative form for the factor analysis model. If we begin at [(1)], then we can write</p>
<p><span class="math display">\[
(y - \mu)(y - \mu)' = (\mathbf{\Lambda} f + \mathbf{u})(\mathbf{\Lambda} f + \mathbf{u})'
\]</span></p>
<p><span class="math display">\[
= \mathbf{\Lambda} f (\mathbf{\Lambda} f)' + \mathbf{u} (\mathbf{\Lambda} f)' + \mathbf{\Lambda} f \mathbf{u}' + \mathbf{u} \mathbf{u}'.
\]</span></p>
<p>Using the three assumptions A1-A3, we can then express the covariance matrix ( ) of <span class="math inline">\(\mathbf{y}\)</span> as</p>
<p><span class="math display">\[
\operatorname{Cov}(\mathbf{y}) = \mathbb{E}(\mathbf{y - \mu)(y - \mu'})
\]</span> <span class="math display">\[
= \mathbf{\Lambda} \mathbb{E}(f f') \mathbf{\Lambda}' + \mathbb{E}(\mathbf{u} f') \mathbf{\Lambda}' + \mathbf{\Lambda} \mathbb{E}(f \mathbf{u}') + \mathbb{E}(\mathbf{u} \mathbf{u}')
\]</span></p>
<p><span class="math display">\[
= \mathbf{\Lambda} \mathbf{\Lambda}' + \mathbf{\Psi}
\tag{F2}
\]</span></p>
<p>demonstrating the equivalence of two formulations of the factor analysis models at [F1] and [F2].</p>
</section>
<section id="communality-and-unique-variance" class="slide level2 scrollable">
<h2>Communality and unique variance</h2>
<ul>
<li>Under all these assumptions, the factor analysis model (F2) implies that the variance of variable <span class="math inline">\(y_i\)</span> can be written as:</li>
</ul>
<p><span class="math display">\[
\sigma_i^2 = \underbrace{\sum_{j=1}^{p} \lambda_{ij}^2}_{\text{communality}} + \overbrace{\psi_i}^{\text{specific variance}}
\]</span></p>
<ul>
<li>So, variance of each observed variable can be split into two parts:</li>
</ul>
<ol type="1">
<li><strong>Communality</strong>, represents the variance shared with the other variables via the common factors.</li>
<li><strong>Specific</strong> or <strong>unique variance</strong> and relates to the variability in <span class="math inline">\(y_i\)</span> not shared with other variables</li>
</ol>
</section>
<section id="covariance" class="slide level2">
<h2>Covariance</h2>
<ul>
<li>The factor model leads to the following expression for the covariance of variables <span class="math inline">\(y_i\)</span> and <span class="math inline">\(y_j\)</span>:</li>
</ul>
<p><span class="math display">\[
\sigma_{ij} = \sum_{l=1}^{p} \lambda_{il}\lambda_{jl}
\]</span></p>
<ul>
<li>The covariances are not dependent on the specific variates in any way; it is the common factors only that aim to account for the relationships between the manifest variables.</li>
</ul>
</section>
<section id="rescaling-the-variables" class="slide level2">
<h2>Rescaling the variables</h2>
<ul>
<li><p>Suppose, we re-scale the variables to <span class="math inline">\(\mathbf{z = C y}\)</span>, where <span class="math inline">\(\mathbf{C} = diag(c_i)\)</span>.</p></li>
<li><p>We can show that the factor analysis model holds for <span class="math inline">\(\mathbf{z}\)</span> with factor loading matrix <span class="math inline">\(\Lambda_z = C \Lambda_y\)</span> and <span class="math inline">\(\Psi_z = C \Psi C = diag(c_i^2 \psi_i)\)</span>.</p></li>
<li><p>The factor loading matrix for the scaled variables <span class="math inline">\(y\)</span> is obtained by multiplying the <span class="math inline">\(i\)</span>th row of <span class="math inline">\(\mathbf{y}\)</span> by <span class="math inline">\(c_i\)</span>, and the same applies to specific variances.<br>
</p></li>
<li><p>If the scaling factors are <span class="math inline">\(c_i = 1/s_i\)</span> (where <span class="math inline">\(s_i\)</span> is the standard deviation of <span class="math inline">\(y_i\)</span>), the factor analysis model effectively applies to the correlation matrix instead of the covariance matrix.<br>
</p></li>
<li><p>As a result, factor analysis can be conducted using either the covariance or correlation matrix, yielding equivalent outcomes.</p></li>
</ul>
</section>
<section id="r-functions" class="slide level2">
<h2><code>R</code> functions</h2>
<ul>
<li><p>In <code>R</code>, we use <code>factanal(x,factors=m)</code> where <code>x</code> is the data arranged in the usual fashion as columns of variables and rows for observations.</p></li>
<li><p>The <code>m</code> parameter specifies how many factors we want <code>R</code> to fit.</p></li>
<li><p>Usually, we try to fit a relatively small number of factors (at most three or four) because of the difficulty of their interpretation.</p></li>
<li><p>There is also a formal statistical hypothesis test performed in factanal() testing whether m factors are adequate or not.</p></li>
</ul>
</section>
<section id="solutions" class="slide level2 scrollable">
<h2>Solutions?</h2>
<p>Two main methods:</p>
<ol type="1">
<li>principal factor analysis, and</li>
<li>maximum likelihood factor analysis,</li>
</ol>
<h3 id="principal-factor-method">Principal Factor method</h3>
<ul>
<li><p>Try to ignore the diagonal elements. Replace with an estimate of the common variance. Usually the squared multiple correlation between one variable and the others is used.</p></li>
<li><p>The method may also be iterated.</p></li>
</ul>
</section>
<section id="maximum-likelihood-method" class="slide level2">
<h2>Maximum Likelihood method</h2>
<ul>
<li><p>fits the model by fitting the covariance/correlation structure using the likelihood function for the multivariate normal distribution.</p></li>
<li><p>The likelihood involves the variance covariance matrix which may be written as a function of the loadings and specific variances.</p></li>
<li><p>Requires iterative solution</p></li>
</ul>
</section>
<section id="hemangioma-example" class="slide level2">
<h2>Hemangioma example</h2>
<ul>
<li><p>Most common childhood cancer diagnosis. It appears as a lump of blood vessels on the skin. If left untreated it usually resolves within a few years. Some parents opt for surgical removal.</p></li>
<li><p>Data: measurements on several genetic markers thought to be related to the disease.</p></li>
<li><p>The age, measured in days, is the infant’s age at the time of the surgery, i.e., how long the infant was exposed to the tumor.</p></li>
<li><p>It was hypothesized exposure to the tumor might influence the magnitude of the expression of some of these genes over time.</p></li>
</ul>
</section>
<section id="hemangioma-data" class="slide level2">
<h2>Hemangioma Data</h2>

<img data-src="hemangioma.png" class="r-stretch quarto-figure-center"><p class="caption">Hemangioma example</p></section>
<section id="hemangioma-example-1" class="slide level2">
<h2>Hemangioma example</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href=""></a>hemangioma <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"hemangioma_data.csv"</span>)</span>
<span id="cb1-2"><a href=""></a>fa.out <span class="ot">&lt;-</span> <span class="fu">factanal</span>(hemangioma, <span class="at">factors =</span> <span class="dv">3</span>)</span>
<span id="cb1-3"><a href=""></a>fa.out<span class="sc">$</span>loadings</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Loadings:
      Factor1 Factor2 Factor3
Age                   -0.168 
RB            -0.170   0.964 
p16    0.381   0.747         
DLK            0.971  -0.211 
Nanog  0.475           0.389 
C.Myc  0.930   0.280  -0.227 
EZH2   0.614  -0.115   0.343 
IGF.2  0.767   0.351   0.187 

               Factor1 Factor2 Factor3
SS loadings      2.225   1.751   1.364
Proportion Var   0.278   0.219   0.170
Cumulative Var   0.278   0.497   0.668</code></pre>
</div>
</div>
<p>We examine the three-factor loadings to see <code>(C.Myc, EZH2, IGF-2)</code> collectively appear to work together in the first factor. The second factor is almost entirely the <code>DLK</code> marker. The third factor is almost represented by the <code>RB</code> marker. <code>Age</code> plays almost no part, contrary to our original hypothesis.</p>
</section>
<section id="uniqueness" class="slide level2">
<h2>Uniqueness</h2>
<ul>
<li>Recall the factor model:</li>
</ul>
<p><span class="math display">\[
\Sigma = \mathbf{\Lambda} \mathbf{\Lambda}' + \mathbf{\Psi}
\tag{F2}
\]</span></p>
<ul>
<li><p>The uniqueness, sometimes referred to as noise, corresponds to the proportion of variability, which can not be explained by a linear combination of the factors. This is the <span class="math inline">\(\Psi\)</span> in (F2).</p></li>
<li><p>A high uniqueness for a variable indicates that the factors do not account well for its variance.</p></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href=""></a>fa.out<span class="sc">$</span>uniquenesses</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Age         RB        p16        DLK      Nanog      C.Myc       EZH2 
0.96312288 0.03388095 0.29067961 0.00500000 0.61640634 0.00500000 0.49146204 
     IGF.2 
0.25418248 </code></pre>
</div>
</div>
</section>
<section id="loadings-again" class="slide level2">
<h2>Loadings, again</h2>
<ul>
<li><p>Loadings range from −1 to 1, this is the <span class="math inline">\(\Lambda\)</span>.</p></li>
<li><p>The loadings are the contribution of each original variable to the factor. Variables with a high loading are well explained by the factor.</p></li>
<li><p>Notice there is no entry for certain variables. That is because R does not print loadings less than 0.1</p></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>
Loadings:
      Factor1 Factor2 Factor3
Age                   -0.168 
RB            -0.170   0.964 
p16    0.381   0.747         
DLK            0.971  -0.211 
Nanog  0.475           0.389 
C.Myc  0.930   0.280  -0.227 
EZH2   0.614  -0.115   0.343 
IGF.2  0.767   0.351   0.187 

               Factor1 Factor2 Factor3
SS loadings      2.225   1.751   1.364
Proportion Var   0.278   0.219   0.170
Cumulative Var   0.278   0.497   0.668</code></pre>
</div>
</div>
</section>
<section id="communality-and-specific-variance" class="slide level2">
<h2>Communality and Specific variance</h2>
<ul>
<li>Recall the decomposition:</li>
</ul>
<p><span class="math display">\[
\sigma_i^2 = \underbrace{\sum_{j=1}^{p} \lambda_{ij}^2}_{\text{communality}} + \overbrace{\psi_i}^{\text{specific variance}} \Rightarrow \quad \sum_{j=1}^{p} \lambda_{ij}^2/\sigma_i^2 + \psi_i/\sigma_i^2 = 1.
\]</span></p>
</section>
<section id="communality-and-specific-variance-1" class="slide level2">
<h2>Communality and Specific variance</h2>
<ul>
<li><p>Here, the <span class="math inline">\(\sum_j \lambda_{ij}^2/\sigma_i^2\)</span> gives the fraction of the variable’s total variance explained by the factor <span class="math inline">\(\sim\)</span> communality.</p></li>
<li><p>We calculate the communality by subtracting the uniquenesses from 1</p></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href=""></a><span class="fu">apply</span>(fa.out<span class="sc">$</span>loadings<span class="sc">^</span><span class="dv">2</span>, <span class="dv">1</span>, sum) <span class="co"># communality</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Age         RB        p16        DLK      Nanog      C.Myc       EZH2 
0.03688578 0.96611902 0.70931820 0.99500266 0.38359679 0.99500472 0.50853609 
     IGF.2 
0.74581358 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href=""></a><span class="dv">1</span> <span class="sc">-</span> <span class="fu">apply</span>(fa.out<span class="sc">$</span>loadings<span class="sc">^</span><span class="dv">2</span>, <span class="dv">1</span>, sum) <span class="co"># uniqueness</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Age          RB         p16         DLK       Nanog       C.Myc 
0.963114225 0.033880978 0.290681797 0.004997344 0.616403209 0.004995279 
       EZH2       IGF.2 
0.491463908 0.254186419 </code></pre>
</div>
</div>
<ul>
<li>An appropriate factor model results in low values for uniqueness and high values for communality.</li>
</ul>
</section>
<section id="proportion-of-variance-explained" class="slide level2">
<h2>Proportion of variance explained</h2>
<ul>
<li>The table below the loadings shows the proportion of variance explained by each factor, with cumulative variance (<code>Cumulative Var</code>) ranging from 0 to 1.<br>
</li>
<li>The row <code>Proportion Var</code> indicates the variance explained by each factor, while <code>SS loadings</code> gives the sum of squared loadings.<br>
</li>
<li>A factor is considered valuable if its <code>SS loadings</code> exceeds 1 (Kaiser’s rule!)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="cell-output cell-output-stdout">
<pre><code>
Loadings:
      Factor1 Factor2 Factor3
Age                   -0.168 
RB            -0.170   0.964 
p16    0.381   0.747         
DLK            0.971  -0.211 
Nanog  0.475           0.389 
C.Myc  0.930   0.280  -0.227 
EZH2   0.614  -0.115   0.343 
IGF.2  0.767   0.351   0.187 

               Factor1 Factor2 Factor3
SS loadings      2.225   1.751   1.364
Proportion Var   0.278   0.219   0.170
Cumulative Var   0.278   0.497   0.668</code></pre>
</div>
</div>
</section>
<section id="testing" class="slide level2">
<h2>Testing</h2>
<ul>
<li><p>The last section of the function output shows the results of a hypothesis test.</p></li>
<li><p><span class="math inline">\(H_0\)</span>: the number of factors in the model, in our case $ m = 3$ factors, is sufficient.</p></li>
<li><p>Usually, we reject <span class="math inline">\(H_0\)</span> if the p-value is less than 0.05, and say that the number of factors is too small.</p></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href=""></a>fa.out<span class="sc">$</span>PVAL</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>objective 
0.3300475 </code></pre>
</div>
</div>
</section>
<section id="p-values" class="slide level2">
<h2>P-values</h2>
<ul>
<li>We can calculate the P-values for a sequence of number of factors, as shown below:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href=""></a><span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(f) <span class="fu">factanal</span>(hemangioma, <span class="at">factors =</span> f, <span class="at">method =</span><span class="st">"mle"</span>)<span class="sc">$</span>PVAL)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  objective   objective   objective   objective 
0.007295322 0.062215517 0.330047536 0.519102928 </code></pre>
</div>
</div>
</section>
<section id="full-output" class="slide level2">
<h2>Full output</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href=""></a>fa.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
factanal(x = hemangioma, factors = 3)

Uniquenesses:
  Age    RB   p16   DLK Nanog C.Myc  EZH2 IGF.2 
0.963 0.034 0.291 0.005 0.616 0.005 0.491 0.254 

Loadings:
      Factor1 Factor2 Factor3
Age                   -0.168 
RB            -0.170   0.964 
p16    0.381   0.747         
DLK            0.971  -0.211 
Nanog  0.475           0.389 
C.Myc  0.930   0.280  -0.227 
EZH2   0.614  -0.115   0.343 
IGF.2  0.767   0.351   0.187 

               Factor1 Factor2 Factor3
SS loadings      2.225   1.751   1.364
Proportion Var   0.278   0.219   0.170
Cumulative Var   0.278   0.497   0.668

Test of the hypothesis that 3 factors are sufficient.
The chi square statistic is 8.03 on 7 degrees of freedom.
The p-value is 0.33 </code></pre>
</div>
</div>
</section>
<section id="rotations" class="slide level2">
<h2>Rotations!</h2>
<ul>
<li><p>Up until now, we have conveniently ignored one problematic feature of the factor analysis model, namely that, there is no unique solution for the factor loading matrix.</p></li>
<li><p>That is, the loadings are only <strong>unique up to a rotation</strong></p></li>
<li><p>That is, you can multiply the factors <span class="math inline">\(f\)</span> by an orthogonal matrix <span class="math inline">\(M_{p \times p}\)</span> without changing the variance decomposition.</p></li>
</ul>

<img data-src="rotation.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="simple-structure-and-rotation" class="slide level2 scrollable">
<h2>Simple structure and rotation</h2>
<ul>
<li><p>The solution from PCA and other extraction methods focus on “fit to the correlation” matrix not on explanation.</p></li>
<li><p>It was observed early that one could rotate the solution (that is rotate the loadings matrix) and not change the fit.</p></li>
<li><p>Hence we look at rotation as a way to make the solution <strong>better</strong>.</p></li>
<li><p>This leads to simple structure as a definition of better.</p></li>
</ul>
<ol type="1">
<li>Most loadings on any factor should be small and only a few should be large in absolute value. A factor only has a few high loading.</li>
<li>A specific row of the loadings matrix <span class="math inline">\(\Lambda\)</span> should display nonzero loadings on only one or at most a few factors. A variable loads on only a few factors</li>
<li>Any pair of factors should have different patterns of loadings. Unique factors.</li>
</ol>
</section></section>
<section id="to-be-continued" class="title-slide slide level1 center">
<h1>To be continued</h1>

</section>

<section>
<section id="section" class="title-slide slide level1 center">
<h1></h1>
<div class="r-fit-text">
<p>Factor Analysis - Part II</p>
</div>
</section>
<section id="recap-model" class="slide level2">
<h2>Recap: Model</h2>
<p>The factor analysis model has two equivalent representation:</p>
<p><span class="math display">\[
\underbrace{\mathbf{y - \mu}}_{p \times 1} = \underbrace{\mathbf{\Lambda}}_{p \times m} ~~ \underbrace{\mathbf{f}}_{m \times 1} + \underbrace{\mathbf{u}}_{p \times 1}. \tag{F1}
\]</span> where</p>
<p><span class="math display">\[
\mathbf{\Lambda} =
\begin{pmatrix}
    \lambda_{11} &amp; \dots &amp; \lambda_{1m} \\
    \vdots &amp; \ddots &amp; \vdots \\
    \lambda_{p1} &amp; \dots &amp; \lambda_{pm}
\end{pmatrix},
\quad
\mathbf{f} =
\begin{pmatrix}
    f_1 \\
    \vdots \\
    f_m
\end{pmatrix},
\quad
\mathbf{u} =
\begin{pmatrix}
    u_1 \\
    \vdots \\
    u_p
\end{pmatrix}.
\]</span></p>
<p>We assume that 1. the random disturbance terms <span class="math inline">\(u_1, \dots, u_p\)</span> are uncorrelated with each other, and 2. uncorrelated with the factors <span class="math inline">\(f_1, \dots, f_m\)</span>.</p>
<p>(The elements of <span class="math inline">\(\mathbf{u}\)</span> are specific to each <span class="math inline">\(y_i\)</span> and hence are generally better known in this context as <em>specific variates</em>.)</p>
</section>
<section id="assumptions-1" class="slide level2">
<h2>Assumptions</h2>
<p>Specifically, we need to assume:</p>
<p><span class="math display">\[
\mathbb{E} f = \mathbf{0} \quad \text{and} \quad \operatorname{Cov}(f) = \mathbb{E}\{f f'\} = \mathbf{I}, \tag{A1}
\]</span></p>
<p><span class="math display">\[
\mathbb{E} \mathbf{u} = \mathbf{0} \quad \text{and} \quad \operatorname{Cov}(\mathbf{u}) = \mathbb{E}(\mathbf{u} \mathbf{u}') = \mathbf{\Psi} = \operatorname{Diag}(\Psi_1, \Psi_2, \dots, \Psi_p).\tag{A2}
\]</span></p>
<p>In addition, the factors ( f ) are uncorrelated with the errors ( u ):</p>
<p><span class="math display">\[
\operatorname{Cov}(f, \mathbf{u}) = \mathbf{0}.\tag{A3}
\]</span></p>
<p>Together, these assumptions (A1-A3) are sufficient for us to fit the model.</p>
</section>
<section id="alternative-formulation" class="slide level2">
<h2>Alternative formulation</h2>
<ul>
<li>Under (A1-A3), we can show that:</li>
</ul>
<p><span class="math display">\[
(y - \mu)(y - \mu)' = (\mathbf{\Lambda} f + \mathbf{u})(\mathbf{\Lambda} f + \mathbf{u})'
\]</span></p>
<p><span class="math display">\[
= \mathbf{\Lambda} f (\mathbf{\Lambda} f)' + \mathbf{u} (\mathbf{\Lambda} f)' + \mathbf{\Lambda} f \mathbf{u}' + \mathbf{u} \mathbf{u}'.
\]</span></p>
<p>Using the three assumptions A1-A3, we can then express the covariance matrix ( ) of <span class="math inline">\(\mathbf{y}\)</span> as</p>
<p><span class="math display">\[
\operatorname{Cov}(\mathbf{y}) = \mathbb{E}(\mathbf{y - \mu)(y - \mu'})
\]</span> <span class="math display">\[
= \mathbf{\Lambda} \mathbb{E}(f f') \mathbf{\Lambda}' + \mathbb{E}(\mathbf{u} f') \mathbf{\Lambda}' + \mathbf{\Lambda} \mathbb{E}(f \mathbf{u}') + \mathbb{E}(\mathbf{u} \mathbf{u}')
\]</span></p>
<p><span class="math display">\[
= \mathbf{\Lambda} \mathbf{\Lambda}' + \mathbf{\Psi}
\tag{F2}
\]</span></p>
</section>
<section id="rotations-1" class="slide level2">
<h2>Rotations!</h2>
<ul>
<li><p>Up until now, we have conveniently ignored one problematic feature of the factor analysis model, namely that, there is no unique solution for the factor loading matrix.</p></li>
<li><p>That is, the loadings are only <strong>unique up to a rotation</strong></p></li>
<li><p>That is, you can multiply the factors <span class="math inline">\(f\)</span> by an orthogonal matrix <span class="math inline">\(M_{p \times p}\)</span> without changing the variance decomposition.</p></li>
</ul>

<img data-src="rotation.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="simple-structure-and-rotation-1" class="slide level2 scrollable">
<h2>Simple structure and rotation</h2>
<ul>
<li><p>The solution from PCA and other extraction methods focus on “fit to the correlation” matrix not on explanation.</p></li>
<li><p>It was observed early that one could rotate the solution (that is rotate the loadings matrix) and not change the fit.</p></li>
<li><p>Hence we look at rotation as a way to make the solution <strong>better</strong>.</p></li>
<li><p>This leads to simple structure as a definition of better.</p></li>
</ul>
<ol type="1">
<li>Most loadings on any factor should be small and only a few should be large in absolute value. A factor only has a few high loading.</li>
<li>A specific row of the loadings matrix <span class="math inline">\(\Lambda\)</span> should display nonzero loadings on only one or at most a few factors. A variable loads on only a few factors</li>
<li>Any pair of factors should have different patterns of loadings. Unique factors.</li>
</ol>
</section>
<section id="rotation" class="slide level2">
<h2>Rotation</h2>
<ul>
<li><p>Most users of factor analysis rotate the solution to try to get factors which display simple structure.</p></li>
<li><p>This makes them easier to interpret.</p></li>
<li><p>Rotation (orthogonal) does not change the distance properties of the loadings.</p></li>
<li><p>Hence total variance and variance explained for each variable remains unchanged.</p></li>
<li><p>However it is now distributed differently amongst the factors.</p></li>
</ul>
</section>
<section id="why-invariant" class="slide level2">
<h2>Why invariant?</h2>
<ul>
<li><p>Suppose we have an orthogonal matrix: <span class="math inline">\(M\)</span> such that <span class="math inline">\(M M^\top = M^\top M = I\)</span>.</p></li>
<li><p>Then, we can modify a factor model by introducing the factor <span class="math inline">\(M\)</span> that won’t change the covariance.</p></li>
</ul>
<p><span class="math display">\[
\mathbf{y = (\Lambda M)(M^\top f) + u}
\]</span></p>
<ul>
<li>New factors: <span class="math inline">\(M^\top f\)</span> and new loadings <span class="math inline">\(\Lambda M\)</span>.</li>
<li>The covariance matrix of the observed:</li>
</ul>
<p><span class="math display">\[
\mathbf{\Sigma = (\Lambda M)(\Lambda M)^\top + \Psi} \\
= \mathbf{\Lambda (M^\top M) \Lambda^\top + \Psi} \\
= \mathbf{\Lambda\Lambda^\top + \Psi} \tag{F2}
\]</span></p>
<ul>
<li>Same covariance structure!</li>
</ul>
</section>
<section id="factor-rotation" class="slide level2">
<h2>Factor rotation</h2>
<ul>
<li><p>It may be that a more interpretable orthogonal solution can be achieved using the equivalent model with loadings <span class="math inline">\(\mathbf{\Lambda^* = \Lambda M}\)</span> for some particular orthogonal matrix, <span class="math inline">\(\mathbf{M}.\)</span></p></li>
<li><p>Such a process is generally known as factor rotation.</p></li>
</ul>
</section>
<section id="criticism-of-factor-analysis" class="slide level2">
<h2>Criticism of Factor Analysis</h2>
<ul>
<li>Factor analysis has faced criticism due to the flexibility of factor rotation.</li>
<li>Some argue that rotation allows researchers to manipulate results to fit preconceived notions (Blackith &amp; Reyment, 1971).</li>
<li>However, these concerns are largely unjustified—rotation is a mathematical tool, not a way to change the underlying structure.</li>
</ul>
</section>
<section id="purpose-of-factor-rotation" class="slide level2">
<h2>Purpose of Factor Rotation</h2>
<ul>
<li>Rotation simplifies factor analysis results, making them more interpretable.</li>
<li>It does <strong>not</strong> alter the fundamental structure of the solution.</li>
<li>A well-rotated solution:
<ul>
<li>Has each variable loading strongly on only <strong>one</strong> factor.</li>
<li>Avoids bipolar factors and intermediate loadings.</li>
<li>Groups variables into distinct sets linked to specific factors.</li>
</ul></li>
</ul>
</section>
<section id="simple-structure-in-factor-analysis" class="slide level2">
<h2>Simple Structure in Factor Analysis</h2>
<p>Thurstone (1931) defined <strong>simple structure</strong> as a desirable property of factor solutions, characterized by:</p>
<ul>
<li>Each row of the factor loading matrix has at least <strong>one zero</strong>.</li>
<li>Each column of the loading matrix has at least <strong>m zeros</strong>.</li>
<li>Every pair of factor columns has variables with <strong>zero loadings in one but not the other</strong>.</li>
<li>With four or more factors, most variables should have <strong>zero loadings in both columns</strong>.</li>
<li>Only a <strong>few variables</strong> should have nonzero loadings in multiple factors.</li>
</ul>
<p>This structure enhances interpretability, making it easier to distinguish between factors.</p>
</section>
<section id="example-of-simple-structure" class="slide level2">
<h2>Example of simple structure</h2>

<img data-src="factor_analysis_files/figure-revealjs/unnamed-chunk-9-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="achieving-simple-structure" class="slide level2 scrollable">
<h2>Achieving Simple Structure</h2>
<ul>
<li><strong>Goal:</strong> Observed variables should form distinct groups, loading <strong>high on one factor</strong> and <strong>low/negligible on others</strong>.<br>
</li>
<li>Medium-sized, ambiguous loadings should be avoided.<br>
</li>
<li>Simple structure is sought <strong>after</strong> determining:
<ol type="1">
<li>The number of common factors.<br>
</li>
<li>The communalities of each variable.</li>
</ol></li>
</ul>
</section>
<section id="rotation-and-factor-independence" class="slide level2">
<h2>Rotation and Factor Independence</h2>
<ul>
<li>Factor loadings are adjusted via <strong>post-multiplication by an orthogonal matrix</strong>.<br>
</li>
<li>This is equivalent to a <strong>rigid rotation</strong> of factor axes.<br>
</li>
<li>Initially, factors are assumed to be <strong>orthogonal (independent)</strong> for simplicity.<br>
</li>
<li>During rotation, we may <strong>relax</strong> this assumption, allowing for <strong>correlated factors</strong> (oblique rotation).</li>
</ul>
</section>
<section id="types-of-rotation" class="slide level2">
<h2>Types of Rotation</h2>
<ul>
<li>Two types of factor rotation:
<ul>
<li><strong>Orthogonal rotation</strong>: Keeps rotated factors <strong>uncorrelated</strong>.</li>
<li><strong>Oblique rotation</strong>: Allows factors to be <strong>correlated</strong>.</li>
</ul></li>
<li><strong>Orthogonal rotation</strong> is achieved by post-multiplying the original loadings matrix by an <strong>orthogonal matrix</strong>.</li>
<li><strong>Oblique rotation</strong> uses a matrix that is <strong>not constrained</strong> to be orthogonal.</li>
</ul>
</section>
<section id="choosing-a-rotation-method" class="slide level2">
<h2>Choosing a Rotation Method</h2>
<ul>
<li><p>The first decision when rotating factors is:<br>
<strong>Should the rotation be orthogonal or oblique?</strong></p></li>
<li><p>There is <strong>no universal answer</strong> to this question.<br>
</p></li>
<li><p>Each method has advantages and disadvantages.</p></li>
<li><p><strong>Oblique rotation</strong> is preferred when the goal is to <strong>best fit the data</strong>.<br>
</p></li>
<li><p><strong>Orthogonal rotation</strong> is preferred when the goal is <strong>generalizability</strong>.</p></li>
</ul>
</section>
<section id="advantages-of-orthogonal-rotation" class="slide level2">
<h2>Advantages of Orthogonal Rotation</h2>
<ul>
<li><strong>Simpler interpretation</strong>:
<ul>
<li>Loadings represent <strong>correlations</strong> between factors and observed variables.<br>
</li>
</ul></li>
<li><strong>Easier to compare across studies</strong>.<br>
</li>
<li><strong>No correlations</strong> between factors, making results more generalizable.</li>
</ul>
</section>
<section id="considerations-for-oblique-rotation" class="slide level2">
<h2>Considerations for Oblique Rotation</h2>
<ul>
<li><strong>Factors can be correlated</strong>, which is <strong>often more realistic</strong> in real-world data.<br>
</li>
<li>However:
<ul>
<li>Loadings <strong>do not directly</strong> represent simple correlations.</li>
<li>Interpretation may be <strong>more complex</strong> due to factor correlations.</li>
</ul></li>
</ul>
</section>
<section id="common-rotation-techniques" class="slide level2">
<h2>Common Rotation Techniques</h2>
<ul>
<li><strong>Orthogonal Rotation</strong>
<ul>
<li>Varimax<br>
</li>
<li>Quartimax</li>
</ul></li>
<li><strong>Oblique Rotation</strong>
<ul>
<li>Oblimin<br>
</li>
<li>Promax</li>
</ul></li>
</ul>
</section>
<section id="varimax-rotation-orthogonal" class="slide level2">
<h2>Varimax Rotation (Orthogonal)</h2>
<ul>
<li>Proposed by <strong>Kaiser (1958)</strong>.<br>
</li>
<li>Aims for <strong>factors with a few large loadings</strong> and as many near-zero loadings as possible.<br>
</li>
<li>Achieved by <strong>iterative maximization</strong> of a quadratic function of the loadings (see <strong>Mardia et al.&nbsp;(1979)</strong>).<br>
</li>
<li>Produces factors that:
<ul>
<li>Have <strong>high correlations</strong> with a <strong>small set of variables</strong>.<br>
</li>
<li>Have <strong>little to no correlation</strong> with other variables.<br>
</li>
</ul></li>
<li><strong>Tends to eliminate</strong> general factors by redistributing factor variance.</li>
</ul>
</section>
<section id="quartimax-rotation-orthogonal" class="slide level2">
<h2>Quartimax Rotation (Orthogonal)</h2>
<ul>
<li>Introduced by <strong>Carroll (1953)</strong>.<br>
</li>
<li>Encourages <strong>each variable</strong> to correlate highly with <strong>one factor</strong> and weakly (or not at all) with others.<br>
</li>
<li>Less commonly used than <strong>varimax</strong>.</li>
</ul>
</section>
<section id="oblimin-rotation-oblique" class="slide level2">
<h2>Oblimin Rotation (Oblique)</h2>
<ul>
<li>Developed by <strong>Jennrich &amp; Sampson (1966)</strong>.<br>
</li>
<li>Seeks <strong>simple structure</strong> in the <strong>factor pattern matrix</strong>.<br>
</li>
<li>Includes a <strong>tuning parameter</strong> to <strong>control the degree of factor correlation</strong>.<br>
</li>
<li><strong>Choosing the parameter value is difficult</strong>, but <strong>Pett, Lackey, &amp; Sullivan (2003)</strong> suggest values between <strong>−0.5 and 0.5</strong> for most applications.</li>
</ul>
</section>
<section id="promax-rotation-oblique" class="slide level2">
<h2>Promax Rotation (Oblique)</h2>
<ul>
<li>Introduced by <strong>Hendrickson &amp; White (1964)</strong>.<br>
</li>
<li>Works by <strong>raising the loadings</strong> from an initial <strong>orthogonal solution</strong> (often varimax) to a power.<br>
</li>
<li>Goal:
<ul>
<li><strong>Best structure</strong> with <strong>lowest power loadings</strong>.<br>
</li>
<li><strong>Minimal correlation</strong> between factors.</li>
</ul></li>
</ul>
</section>
<section id="varimax" class="slide level2">
<h2>Varimax</h2>
<ul>
<li>The default rotation is “varimax” in <code>factanal()</code> function which is one of the orthogonal rotation methods.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href=""></a>heme_varimax <span class="ot">&lt;-</span> <span class="fu">factanal</span>(hemangioma, <span class="at">factors =</span> <span class="dv">3</span>, <span class="at">rotation =</span> <span class="st">"varimax"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="factor_analysis_files/figure-revealjs/unnamed-chunk-11-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="no-rotation" class="slide level2">
<h2>No rotation</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href=""></a>heme_none <span class="ot">&lt;-</span> <span class="fu">factanal</span>(hemangioma, <span class="at">factors =</span> <span class="dv">3</span>, <span class="at">rotation =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="factor_analysis_files/figure-revealjs/unnamed-chunk-13-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="promax-rotation" class="slide level2">
<h2>Promax rotation</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href=""></a>heme_promax <span class="ot">&lt;-</span> <span class="fu">factanal</span>(hemangioma, <span class="at">factors =</span> <span class="dv">3</span>, <span class="at">rotation =</span> <span class="st">"promax"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="factor_analysis_files/figure-revealjs/unnamed-chunk-15-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="other-r-packages" class="slide level2">
<h2>Other <code>R</code> packages</h2>
<ul>
<li><code>psych</code> is a very nice package with lots of tools and functions. The following code does factor analysis for the hemangioma data with varimax rotation.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href=""></a><span class="fu">library</span>(psych)</span>
<span id="cb20-2"><a href=""></a>heme_psych_fa <span class="ot">&lt;-</span> <span class="fu">fa</span>(hemangioma, <span class="co"># input data</span></span>
<span id="cb20-3"><a href=""></a>              <span class="at">nfactors =</span> <span class="dv">3</span>, <span class="co"># number of factors</span></span>
<span id="cb20-4"><a href=""></a>              <span class="at">rotate =</span> <span class="st">"varimax"</span>) <span class="co"># rotation</span></span>
<span id="cb20-5"><a href=""></a>              </span>
<span id="cb20-6"><a href=""></a>heme_psych_fa<span class="sc">$</span>loadings <span class="co"># factor loadings</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Loadings:
      MR1    MR2    MR3   
Age                 -0.735
RB     0.103 -0.441  0.248
p16    0.405  0.700  0.229
DLK           0.865       
Nanog  0.611 -0.282  0.134
C.Myc  0.721  0.381       
EZH2   0.631 -0.242  0.551
IGF.2  0.967  0.215 -0.125

                 MR1   MR2   MR3
SS loadings    2.402 1.762 0.995
Proportion Var 0.300 0.220 0.124
Cumulative Var 0.300 0.521 0.645</code></pre>
</div>
</div>
</section>
<section id="diagrams" class="slide level2">
<h2>Diagrams</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href=""></a><span class="fu">library</span>(psych)</span>
<span id="cb22-2"><a href=""></a>heme_loads <span class="ot">&lt;-</span> heme_promax<span class="sc">$</span>loadings</span>
<span id="cb22-3"><a href=""></a><span class="fu">fa.diagram</span>(heme_loads)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="factor_analysis_files/figure-revealjs/unnamed-chunk-17-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="the-psych-package" class="slide level2">
<h2>The <code>psych</code> package</h2>
<ul>
<li>Supports more functionality. For example, the <code>rotate</code> argument has the following description if you look at the help for <code>fa</code>.</li>
</ul>
<blockquote>
<p>“none”, “varimax”, “quartimax”, “bentlerT”, “equamax”, “varimin”, “geominT” and “bifactor” are orthogonal rotations. “Promax”, “promax”, “oblimin”, “simplimax”, “bentlerQ”, “geominQ” and “biquartimin” and “cluster” are possible oblique transformations of the solution. The default is to do a oblimin transformation, although versions prior to 2009 defaulted to varimax.</p>
</blockquote>
</section>
<section id="other" class="slide level2">
<h2>Other</h2>
<ul>
<li><code>GPArotation</code></li>
</ul>
<blockquote>
<p><code>GPArotation package contains functions for the rotation of factor loadings matrices. The functions implement Gradient Projection (GP) algorithms for orthogonal and oblique rotation. Additionally, a number of rotation criteria are provided. The GP algorithms minimize the rotation criterion function, and provide the corresponding rotation matrix. For oblique rotation, the covariance / correlation matrix of the factors is also provided.</code></p>
</blockquote>
</section>
<section id="some-assessments" class="slide level2">
<h2>SOME ASSESSMENTS</h2>
<ul>
<li><p><strong>Error matrix</strong> - Compute the estimated variance covariance matrix then compare it with the original matrix or compute the residual matrix. Like to have the off diagonals small</p></li>
<li><p><strong>Partial correlation matrix</strong> - correlation matrix between the variables after partialling out the information provided by the factors. For example, the correlation between y1 and y2 adjusted for factor 1. These values should be near zero. (There is a test Kaiser-Meyer-Olkin test for adequacy using the partial correlations).</p></li>
</ul>
</section>
<section id="kmo-test" class="slide level2">
<h2>KMO test</h2>
<p>Henry Kaiser (1970) introduced an Measure of Sampling Adequacy (MSA) of factor analytic data matrices.</p>
<p>“In his delightfully flamboyant style, Kaiser (1975) suggested that KMO &gt; .9 were marvelous, in the .80s, mertitourious, in the .70s, middling, in the .60s, medicore, in the 50s, miserable, and less than .5, unacceptable.”</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href=""></a><span class="fu">KMO</span>(hemangioma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Kaiser-Meyer-Olkin factor adequacy
Call: KMO(r = hemangioma)
Overall MSA =  0.51
MSA for each item = 
  Age    RB   p16   DLK Nanog C.Myc  EZH2 IGF.2 
 0.30  0.27  0.56  0.43  0.84  0.48  0.72  0.55 </code></pre>
</div>
</div>
</section>
<section id="terminology-note" class="slide level2">
<h2>Terminology note</h2>
<ul>
<li><p><strong>Loadings</strong> – associated with correlations with variables and weights, usually applied in orthogonal factor model.</p></li>
<li><p><strong>Pattern</strong> – estimates of weights in factor analysis – analogous to the coefficients in regression (<span class="math inline">\(\Lambda\)</span> in the model) (preferred over structure coeffs). Replaces the loadings matrix in oblique rotation. Squares represent unique contribution of factor to variable.</p></li>
<li><p><strong>Factor Structure</strong> – correlation between the factors and variables, more commonly applied in oblique situations (same as pattern in orthogonal but not oblique due to correlation between factors). Represent unique contribution from factor and contribution from correlation with other factors.</p></li>
<li><p>Pattern usually reported but may not be good when there are highly correlated factors. Analogous to reporting <span class="math inline">\(\beta(x_i)\)</span> in regression versus <span class="math inline">\(corr(y,x_i)\)</span></p></li>
</ul>
</section>
<section id="comments" class="slide level2">
<h2>Comments</h2>
<p>Different methods may yield considerably different solutions and rotated solutions.</p>
<p>Rotation may obscure a dominant factor and is influenced by the number of factors retained.</p>
<p>Rotation should follow the maximum likelihood solution (as the constraint was added to insure uniqueness of a solution).</p>
<p>With PCA, rotation destroys the variance maximizing property.</p>
<p>The total variance explained becomes spread over the factors. For example, factor 1 may not have the largest variance.</p>
</section>
<section id="factor-analysis-vs.-pca" class="slide level2">
<h2>Factor Analysis vs.&nbsp;PCA</h2>
<ul>
<li>Both <strong>reduce dimensionality</strong>, but differ in approach:
<ul>
<li><strong>Factor Analysis</strong> explains <strong>covariances/correlations</strong> using <strong>common factors</strong>.<br>
</li>
<li><strong>PCA</strong> explains <strong>variance</strong> in observed variables.<br>
</li>
</ul></li>
<li>Adding a new <strong>PCA component</strong> (<strong>m → m+1</strong>) leaves prior components unchanged, but adding a factor in <strong>Factor Analysis</strong> can alter all factors.<br>
</li>
<li><strong>PCA scores</strong> are straightforward to compute; <strong>Factor scores</strong> require more complex methods.</li>
</ul>
</section>
<section id="covariance-correlation-and-interpretation" class="slide level2">
<h2>Covariance, Correlation, and Interpretation</h2>
<ul>
<li><strong>PCA:</strong> Principal components from <strong>correlation vs.&nbsp;covariance matrices</strong> are generally <strong>different</strong>.<br>
</li>
<li><strong>Factor Analysis (ML):</strong> Results from <strong>correlation vs.&nbsp;covariance matrices</strong> are <strong>essentially equivalent</strong>.<br>
</li>
<li><strong>Rotation in Factor Analysis</strong> helps interpretability but does not change the mathematical structure.</li>
</ul>
</section>
<section id="comments---sample-size" class="slide level2">
<h2>COMMENTS - SAMPLE SIZE</h2>
<ul>
<li><p>No hard rules</p></li>
<li><p>but with good factors</p>
<ul>
<li>at least 5 observations per variable</li>
<li>at least 100 observations</li>
</ul></li>
<li><p>More observations required for weaker relationships <span class="math inline">\(n &gt; 10\times p\)</span>.</p></li>
</ul>
</section>
<section id="comments-strategies" class="slide level2">
<h2>COMMENTS/ STRATEGIES</h2>
<ul>
<li><p>Try more than one method to see if results are consistent!</p></li>
<li><p>If the data is large enough, try splitting the data in half to see if the solutions are similar.</p></li>
<li><p>The choice of the number of factors is critical when the solution is rotated. Try dropping or adding a factor to see how stable the rotated solution is.</p></li>
<li><p>Plot the data if possible – look for linearity and normality Remove variables that are linear combinations of other variables.</p></li>
<li><p>These cause problems with ML methods Remove variables that represent single factors (factors that are made of only one variable).</p></li>
<li><p>These represent solitary factors that may confuse rotation methods. You can of course look at these but worry about the effect of keeping them</p></li>
</ul>
</section>
<section id="envoi" class="slide level2">
<h2>Envoi</h2>
<p><img data-src="summary_everitt.png" width="822"></p>
<div class="quarto-auto-generated-content">
<p><img src="vt.png" class="slide-logo"></p>
<div class="footer footer-default">

</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="factor_analysis_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="factor_analysis_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="factor_analysis_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="factor_analysis_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="factor_analysis_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="factor_analysis_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="factor_analysis_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="factor_analysis_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="factor_analysis_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="factor_analysis_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    

    <script>

      // htmlwidgets need to know to resize themselves when slides are shown/hidden.

      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current

      // slide changes (different for each slide format).

      (function () {

        // dispatch for htmlwidgets

        function fireSlideEnter() {

          const event = window.document.createEvent("Event");

          event.initEvent("slideenter", true, true);

          window.document.dispatchEvent(event);

        }

    

        function fireSlideChanged(previousSlide, currentSlide) {

          fireSlideEnter();

    

          // dispatch for shiny

          if (window.jQuery) {

            if (previousSlide) {

              window.jQuery(previousSlide).trigger("hidden");

            }

            if (currentSlide) {

              window.jQuery(currentSlide).trigger("shown");

            }

          }

        }

    

        // hookup for slidy

        if (window.w3c_slidy) {

          window.w3c_slidy.add_observer(function (slide_num) {

            // slide_num starts at position 1

            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);

          });

        }

    

      })();

    </script>

    

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        // For code content inside modals, clipBoardJS needs to be initialized with a container option
        // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>